### PR TITLE
[12.0][ADD] account_move_multi_company

### PR DESCRIPTION
**Pull request by Khalid-SerpentCS**
Odoo allows a journal entry to be posted in the same company. In a multi-company environment, it allows to have inter-company transactions at the sale order and purchase order level.

Some group/holding have setup a company structure where an employee is paid by a company of the group but is working part-time for another company of the group.

When the accounting team is loading the accounting entry for the payroll, the amount dedicated to the other companies needs to be transferred to those companies.

#183

### CODE REVIEW

File: account_move_multi_company/models/account_move.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        # Override to prevent ValidationError
+        # Method in odoo/addons/account/models/account_move line 357
+        return True
+
+
+class AccountMoveLine(models.Model):
+    _inherit = "account.move.line"
+
+    transfer_to_company_id = fields.Many2one(
+        'res.company', string='Transfer to')
```
Comment (max3903) [inline account_move_multi_company/models/account_move.py]: @Khalid-SerpentCS Can you create one file per model?

Ref:
https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#file-naming

File: account_move_multi_company/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    'version': '12.0.1.0.0',
+    'category': 'Accounting & Finance',
+    'author': 'Open Source Integrators, Odoo Community Association (OCA)',
+    'website': 'https://github.com/OCA/multi-company',
+    'license': 'AGPL-3',
+    'depends': [
+        'account_invoice_consolidated',
+    ],
+    'data': [
+        'views/account_move_view.xml',
```
Comment (max3903) [inline account_move_multi_company/__manifest__.py]: @Khalid-SerpentCS We need to add a file `security/ir_rule.xml` to allow users in the billing group to read all companies.

File: account_move_multi_company/models/account_move.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+                # Journal Entry for each dedicated company
+                dedicated_company_move = self.env['account.move'].sudo()
+                for transfer_to_company_id in dedicated_companies_vals:
+                    dedicated_company_move += self.env['account.move'].sudo().\
+                        create(dedicated_companies_vals[
+                               transfer_to_company_id])
+                self += dedicated_company_move
+
+        return super(AccountMove, self).post()
```
Comment (max3903) [inline account_move_multi_company/models/account_move.py]: @Khalid-SerpentCS Can you create additional methods to reduce the size of this one and allow easier extension?

As suggested in the design, I think we need to call super() first to be able to reconcile the entries.

File: account_move_multi_company/tests/test_account_move.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+        company_two_due_tofrom_moves_before =\
+            self.env['account.move'].sudo().search_count(
+                [('journal_id', '=',
+                    self.due_to_due_from_journal_company_two.id),
+                 ('company_id', '=', self.company_two.id)])
+
+        self.payroll_move.post()
+
+        # Main company Due To/Due From Move After Payroll Journal Entry Post
+        main_company_due_tofrom_moves__after =\
```
Comment (max3903) [inline account_move_multi_company/tests/test_account_move.py]:
```suggestion
        main_company_due_tofrom_moves_after =\
```

File: account_move_multi_company/tests/test_account_move.py [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+        # Post
+
+        company_two_due_tofrom_moves_after =\
+            self.env['account.move'].sudo().search_count(
+                [('journal_id', '=',
+                    self.due_to_due_from_journal_company_two.id),
+                 ('company_id', '=', self.company_two.id)])
+
+        # Check for Company One
+        self.assertEqual(main_company_due_tofrom_moves__after,
```
Comment (max3903) [inline account_move_multi_company/tests/test_account_move.py]:
```suggestion
        self.assertEqual(main_company_due_tofrom_moves_after,
```

**Comment by max3903** [comment]
@Khalid-SerpentCS Please review and merge https://github.com/SerpentConsultingServices/multi-company/pull/1

**Comment by Khalid-SerpentCS** [comment]
@max3903  UT is failed because of record rule. https://travis-ci.org/OCA/multi-company/jobs/637816417#L1398

**Comment by max3903** [comment]
@Khalid-SerpentCS By default, a new user belongs to the Billing group. Can you update the test (in this pull request) to add the user in the Employee group only?

Thanks!

CC: @pedrobaeza

**Comment by Khalid-SerpentCS** [comment]
@max3903 , I have tested with the user with Billing right. But UT is not passing.

**Comment by max3903** [comment]
@Khalid-SerpentCS @osi-scampbell @agyamuta Please review https://github.com/SerpentConsultingServices/multi-company/pull/2

**Comment by max3903** [comment]
@Khalid-SerpentCS https://github.com/SerpentConsultingServices/multi-company/pull/3

**Review by max3903 — APPROVED** [review]

**Review by murtuzasaleh — APPROVED** [review]

**Pull request merged**